### PR TITLE
FALCON-1956 Graphite Plugin for monitoring

### PR DESCRIPTION
**Pull request by PraveenAdlakha**

### CODE REVIEW

**Comment by vrangan** [comment]
+1 for Graphite support.

As a follow up, We probably need to see how we can support different metrics providers.   I am primarily thinking about AMS, as it is mostly available for most of the Falcon users (as it is part of Ambari) and integrated with rest of the Hadoop metrics and now with Grafana dashboards

File: prism/src/main/java/org/apache/falcon/plugin/GraphiteNotificationPlugin.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import org.joda.time.DateTime;
+import org.joda.time.Seconds;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Graphite Notification Plugin.
+ */
+public class GraphiteNotificationPlugin implements MonitoringPlugin {
```
Comment (sriksun) [inline prism/src/main/java/org/apache/falcon/plugin/GraphiteNotificationPlugin.java]: Please do consider Adding an AlertingPlugin additionally to capture permanent failures as well.

**Comment by sriksun** [comment]
:+1:

File: prism/src/main/java/org/apache/falcon/plugin/GraphiteNotificationPlugin.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            String prefix = StartupProperties.get().getProperty("falcon.graphite.prefix");
+            if (entityType.equals(EntityType.PROCESS.name())) {
+                Entity entity = ConfigurationStore.get().get(EntityType.PROCESS, entityName);
+                Process process = (Process) entity;
+                String pipeline =  StringUtils.isNotBlank(process.getPipelines()) ? process.getPipelines() : "default";
+
+
+                if ((message.getAction().equals("wf-instance-succeeded"))) {
+                    Long timeTaken =  message.getExecutionTime() / 1000000000;
+                    String metricsName = prefix + pipeline + ".GENERATE." + entityName + ".processing_time";
```
Comment (pavankumar526) [inline prism/src/main/java/org/apache/falcon/plugin/GraphiteNotificationPlugin.java]: All time metrics are in secs ? We should be consistent in this

File: prism/src/main/java/org/apache/falcon/plugin/GraphiteNotificationPlugin.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                    metricNotificationService.publish(metricsName, timeTaken);
+
+                    DateTime nominalTime = new DateTime(message.getDimensions().get("nominal-time"));
+                    DateTime startTime = new DateTime(message.getDimensions().get("start-time"));
+                    metricsName = prefix + pipeline + ".GENERATE." + entityName + ".start_delay";
```
Comment (pavankumar526) [inline prism/src/main/java/org/apache/falcon/plugin/GraphiteNotificationPlugin.java]: Why GENERATE this is added ?

**Comment by pavankumar526** [comment]
Looks good otherwise

File: prism/src/main/java/org/apache/falcon/plugin/GraphiteNotificationPlugin.java (hunk 2 above)
Comment (pavankumar526) [inline prism/src/main/java/org/apache/falcon/plugin/GraphiteNotificationPlugin.java]: Add cluster name as we discussed.

**Comment by pavankumar526** [comment]
+1 looks good

**Pull request closed without merging**